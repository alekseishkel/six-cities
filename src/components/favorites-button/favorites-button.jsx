import React, {Component} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import ActionCreator from '../../action-creator/action-creator';

class FavoritesButton extends Component {
  constructor() {
    super();

    this.state = {
      isActive: false
    };
  }

  componentDidMount() {
    if (this.props.cardInfo.is_favorite === true) {
      this.setState({isActive: true});
    }
  }

  render() {
    const {isActive} = this.state;
    const {cardInfo, onFavoritesButtonAddClick, onFavoritesButtonRemoveClick} = this.props;

    let isActiveModifier = null;

    if (isActive === true) {
      isActiveModifier = `--active`;
    }

    return (
      <button
        className={`place-card__bookmark-button place-card__bookmark-button${isActiveModifier} button`}
        type="button"
        onClick={() => {
          cardInfo.is_favorite = !isActive;

          this.setState((state) => {
            return {
              isActive: !state.isActive
            };
          });

          if (isActive === false) {
            onFavoritesButtonAddClick(cardInfo);
          }

          if (isActive === true) {
            onFavoritesButtonRemoveClick(cardInfo.id);
          }
          console.log(cardInfo);
        }}>
        <svg className="place-card__bookmark-icon" width="18" height="19">
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">In bookmarks</span>
      </button>
    );
  }
}

FavoritesButton.propTypes = {
  cardInfo: PropTypes.object.isRequired,
  // onFavoritesButtonClick: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onFavoritesButtonAddClick: (offer) => {
    dispatch(ActionCreator.addOfferToFavorites(offer));
  },
  onFavoritesButtonRemoveClick: (offer) => {
    dispatch(ActionCreator.removeOfferFromFavorites(offer));
  },
});

export {FavoritesButton};

export default connect(null, mapDispatchToProps)(FavoritesButton);
