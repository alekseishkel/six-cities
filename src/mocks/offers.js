const offersCardsInfo = [{
  description: `Beautiful & luxurious apartment at great location`,
  img: `img/apartment-01.jpg`,
  price: `€120`,
  type: `Apartment`,
  isPremium: true
},
{
  description: `Wood and stone place`,
  img: `img/room.jpg`,
  price: `€80`,
  type: `Private room`,
  isPremium: false
},
{
  description: `Canal View Prinsengracht`,
  img: `img/apartment-02.jpg`,
  price: `€132`,
  type: `Apartment`,
  isPremium: false
},
{
  description: `Nice, cozy, warm big bed apartment`,
  img: `img/apartment-03.jpg`,
  price: `€180`,
  type: `Apartment`,
  isPremium: false
}];

export default offersCardsInfo;
