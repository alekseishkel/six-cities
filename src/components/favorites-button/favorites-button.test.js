import React from 'react';
import renderer from 'react-test-renderer';

import {FavoritesButton} from './favorites-button.jsx';

import {createBrowserHistory as history} from 'history';

import offers from '../../mocks/offers';

it(`FavoritesButton correctly renders after relaunch`, () => {

  const tree = renderer.create(
      <FavoritesButton
        cardInfo={offers[0]}
        history={history()}
        isAuthorized={true}
        offers={offers}
        onFavoritesButtonAddClick={jest.fn()}
        onFavoritesButtonRemoveClick={jest.fn()}
        favorites={[offers[0], offers[2], offers[4]]}
        type={`property`}
      />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
