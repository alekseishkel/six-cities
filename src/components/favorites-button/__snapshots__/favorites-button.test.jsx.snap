// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesButton correctly renders after relaunch 1`] = `
<button
  className="property__bookmark-button place-card__bookmark-buttonnull button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={33}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In bookmarks
  </span>
</button>
`;
